src/client.rs

```diff
@@ -47,21 +49,134 @@ pub struct Client {
     inner: Arc<ClientRef>,
 }
 
-impl Client {
-    /// Constructs a new `Client`.
-    pub fn new() -> ::Result<Client> {
-        let mut client = try_!(new_hyper_client());
-        client.set_redirect_policy(::hyper::client::RedirectPolicy::FollowNone);
+/// Represent an X509 certificate.
+pub struct Certificate(hyper_native_tls::Certificate);
+
+impl Certificate {
+    /// Create a `Certificate` from a binary DER encoded certificate
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// # use std::io::Read;
+    /// # fn main() {
+    /// let mut buf = Vec::new();
+    /// let _ = ::std::fs::File::open("my-cert.der").unwrap().read_to_end(&mut buf).unwrap();
+    /// let cert = reqwest::Certificate::from_der(&buf).unwrap();
+    /// # }
+    /// ```
+    pub fn from_der(der: &[u8]) -> ::Result<Certificate> {
+        let inner = try_!(
+            hyper_native_tls::Certificate::from_der(der)
+                .map_err(|e| ::hyper::Error::Ssl(Box::new(e))));
+        Ok(Certificate(inner))
+    }
+}
+
+/// A `TlsClient` can be used to create a `Client` with a custom TLS configuration:
+///
+/// - with hostname verification disabled
+/// - with one or multiple custom certificates
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// extern crate reqwest;
+///
+/// use std::fs::File;
+/// use std::io::Read;
+///
+/// use reqwest::{Certificate, TlsClient};
+///
+/// fn main() {
+///     // get a client builder
+///     let mut tls_client = TlsClient::new().unwrap();
+///
+///     // read a local binary DER encoded certificate
+///     let mut buf = Vec::new();
+///     let _ = File::open("my-cert.der").unwrap().read_to_end(&mut buf).unwrap();
+///
+///     // create a certificate
+///     let cert = Certificate::from_der(&buf).unwrap();
+///
+///     // add the certificate
+///     tls_client.add_root_certificate(cert).unwrap();
+///
+///     // create the actual client
+///     let client = tls_client.into_client().unwrap();
+///
+///     // [...]
+/// }
+/// ```
+///
+pub struct TlsClient {
```

Comment:
I actually liked that naming you have before, `ClientBuilder`. We can eventually move several of the config properties to the builder and remove the need for the `Mutex` and atomics...